File: python/3_advanced_data_strutures/trees/1_basic.py
from typing import Any, List


class TreeNode:
    def __init__(self, data: Any):
        self.data = data
        self.children: List["TreeNode"] = []

    def add_child(self, child: "TreeNode"):
        self.children.append(child)


# To print
def print_tree(root: TreeNode, level: int = 0):
    print("    " * level + f"- {root.data}")
    for child in root.children:
        print_tree(child, level + 1)


# Build tree
root = TreeNode("CEO")

head_of_sales = TreeNode("Head of Sales")
head_of_marketing = TreeNode("Head of Marketing")
sales_executive = TreeNode("Sales Executive")
marketing_executive = TreeNode("Marketing Executive")

root.add_child(head_of_sales)
root.add_child(head_of_marketing)

head_of_sales.add_child(sales_executive)
head_of_marketing.add_child(marketing_executive)

# Print tree
print_tree(root)

File: go/3_advancedDataStructures/1_trees/1_basic.go
package trees

type BasicTreeNode[T any] struct {
	Value    T
	Children []BasicTreeNode[T]
}

func (b *BasicTreeNode[T]) AddChild(child BasicTreeNode[T]) {
	b.Children = append(b.Children, child)
}

func BuildTree() BasicTreeNode[string] {
	root := BasicTreeNode[string]{Value: "CEO"}

	headOfSales := BasicTreeNode[string]{Value: "Head of Sales"}
	headOfMarketing := BasicTreeNode[string]{Value: "Head of Marketing"}
	salesExec := BasicTreeNode[string]{Value: "Sales Executive"}
	marketingExec := BasicTreeNode[string]{Value: "Marketing Executive"}

	headOfSales.AddChild(salesExec)
	headOfMarketing.AddChild(marketingExec)

	root.AddChild(headOfSales)
	root.AddChild(headOfMarketing)

	return root
}
